### DISCUSSION TITLE
Foreground notification handling (onMessage) and NotificationServiceExtension

### DISCUSSION THREAD

**Discussion by bumpingChris** [invertase/react-native-firebase#6556] [Q&A]
Hi,

In iOS, I use NotificationServiceExtension to modify the message before showing to the user. This works fine when the app is in the background. But when in the foreground, I see that the NotificationServiceExtension does get called when the message arrives, BUT the onMessage listener does not see the modified message. 

How can I get onMessage to see the modified message?

Thank you for your help.

**Comment by mikehardy** [comment]
Interesting - I'm not sure, I have not considered that use case, it's literally never come up before :-)

There may not be a way owing to control flow / when things happen, and what data is provided to onMessage.

More concretely: onMessage is a hook called in our code from firebase-ios-sdk and we are just like a postal carrier, delivering the message to higher levels that we received from lower levels

There *may* be something you can do with regard to swizzling such that the call chain has your extension prior to the firebase-ios-sdk Messaging delegate swizzle (if I'm understanding that correctly) - but that would be experimental on your behalf, you would have to play around with it to see if a) I am even guessing in the correct area b) you even have control of the order and c) you can effectively change it to work the way you want

A separate path may be to use NSUserDefault in the extension to leave yourself a clue as to what happened then in react-native check it (via https://reactnative.dev/docs/settings) - a bit less elegant then a perfect swizzle+ordering thing but also a lot less subtle.

Maybe useful?

**Reply by bumpingChris** [reply]
@mikehardy Thank you, Mike, for your thorough answer. I was hoping, of course, that there would be an easy way to get the result from the extension, but if there isn't I could recode what the extension does in the onMessage listener. Not as elegant for sure, but much easier than experimenting with swizzling.

Thanks again.
Chris

**Reply by mikehardy** [reply]
Luckily at least NSUserDefaults is really easy to work with both in native code and react-native level, it's a good escape hatch. Good luck with your project!
